lib/Url/UrlManager.php

```diff
@@ -281,12 +281,16 @@ public static function resolveUrl(Url $url)
             return null;
         }
 
-        $rewriterSuffix = Url::getRewriter()->getSuffix();
-        if (\rex::isFrontend() && $rewriterSuffix && substr($url->getPath(), -strlen($rewriterSuffix)) !== $rewriterSuffix) {
-            header('HTTP/1.1 301 Moved Permanently');
-            header('Location: '.$url->getPath().$rewriterSuffix.$url->getQuery());
-            exit;
-        }
+	// Weiterleitung auf URL mit Suffix, wenn Suffix fehlt
+	$rewriterSuffix = Url::getRewriter()->getSuffix();
+	if ($resolve && \rex::isFrontend() && $rewriterSuffix && substr($url->getPath(), -strlen($rewriterSuffix)) !== $rewriterSuffix) {
+		$url->handleRewriterSuffix();
+		if(count(UrlManagerSql::getByUrl(new Url($url->__toString()))) == 1) {
```

Comment:
Müsste das nicht auch so funktionieren?

```suggestion
		if(count(UrlManagerSql::getByUrl($url))) == 1) {
```

Reply:
Wenn ich das nicht mache, werden die Parameter in der Weiterleitung nicht berücksichtigt.

Reply:
Das scheint mir dennoch falsch zu sein. Denn in der getByUrl Methode wir die URL via __toString gesetzt.

Reply:
Irgendwo stimmt was noch nicht. Wenn ich die URL https://test.kaltenbach.com/en/circular-sawing-machines?test=1 eingebe wird mit meinem Code auf https://test.kaltenbach.com/en/circular-sawing-machines/?test=1 weitergeleitet. Ändere ich den Code auf deinen Vorschlag wird auf https://test.kaltenbach.com/en/circular-sawing-machines/ weitergeleitet. Woran liegt das?

Reply:
Ah, hab den Unterschied. Beim "alten" URL Objekt hat der Parameter `handleRewriterSuffix` den Wert `true`. Beim "neuen" URL Objekt ist der Suffix schon in der URL hinzugefügt und der Parameter `handleRewriterSuffix` hat den Wert `false`. Darin liegt der Unterschied.

Reply:
Ist das korrekt, dass die Methode` __toString` die Url-Params nicht mehr mit ausgibt sobald `handleRewriterSuffix` auf `true` steht?

Reply:
Also wenn du mich fragst: nein, das ist nicht richtig. Ich schließe aus deiner Frage, dass das ein Bug ist, oder?

Reply:
Das Problem liegt nicht in der `__toString()` Methode, sondern hier: https://github.com/tbaddade/redaxo_url/blob/a2db315c537210d39beb4269a5548ccc7679655c/lib/Url/UrlManagerSql.php#L283-L284. Diese beiden Zeilen kürzen den relevanten Code aus dem `$url` Objekt raus. Durch das erstellen des "neuen" URL Objektes wird das vermieden.

Reply:
> Diese beiden Zeilen kürzen den relevanten Code aus dem $url Objekt raus. Durch das erstellen des "neuen" URL Objektes wird das vermieden.

Verstehe ich nicht … Das sollte bei beiden hier passieren. Egal ob du ein neues Objekt übergibst oder das ältere. Haben wir hier beide irgendwelche Denkfehler?

Reply:
Schau dir mal den Screenshot an.
![Bildschirmfoto von 2020-04-22 11-59-03](https://user-images.githubusercontent.com/4131072/79968689-c4322880-8490-11ea-9255-d3a4c1597f55.png)
Was siehst du dort? Den Dump des `$url` Objektes und danach von `$url->__toString()` vor der Zeile die du als Änderung vorschlägst und den Dump des `$url` Objektes und danach von `$url->__toString()` nach der Zeile die du als Änderung vorschlägst.
Verstehst du, was ich meine? Beim Aufruf von `UrlManagerSql::getByUrl($url)` wird das `$url` Objekt verändert.

Reply:
Kannst du bitte noch über den Screenshot kur schreiben, welches Objekt das ist. Alt und Neu oder so etwas. Sodass wir später es schneller erfassen können und nicht noch einmal alle texte lesen müssen ;) … Danke

Reply:
Klaro. Hier ein ausführlicherer Screenshot:
![Bildschirmfoto von 2020-04-22 12-55-37](https://user-images.githubusercontent.com/4131072/79973933-af599300-8498-11ea-82b8-b059388c31a5.png)

Reply:
Ich habe mal meine unnötigen Kommentare gelöscht. So ist es leichter zu lesen. Vielleicht willst du auch noch den einen oder anderen Kommentar entfernen.

Was denkst du: liegt hier ein neuer Bug in der Klasse `UrlManagerSql` vor oder ist mein Code genau richtig um mit der Situation umzugehen?

Reply:
Dein Code birgt für mich Potenzial, um später wieder darüber zu stolpern. Am sinnvollsten wäre es wohl in der `UrlManagerSql` das $url Objekt zunächst zu clonen. Dazu müsste man dann aber alle bisherigen `UrlManagerSql` Aufrufe nochmals gegen testen.

Reply:
OK. Objekt wird nun [geklont](https://github.com/tbaddade/redaxo_url/pull/168/commits/2642b65d9bcc010d07f2dccd04e838e4e65661be). Das hat keinerlei Auswirkungen auf das Addon, da die Methode `getByUrl()` nur an zwei Stellen ([hier im PR](https://github.com/TobiasKrais/redaxo_url/blob/22d50d57fbd1c013b56d5c9574f6929b2578d0ec/lib/Url/UrlManager.php#L289) und [hier im Addon](https://github.com/tbaddade/redaxo_url/blob/a2db315c537210d39beb4269a5548ccc7679655c/lib/Url/UrlManager.php#L291)) verwendet wird die das `$url` Objekt nachher nicht weiter verarbeiten.